mapboxgl/layers.py

```diff
@@ -0,0 +1,745 @@
+        self.radius_stops = radius_stops
+        self.color_function_type = color_function_type
+        self.color_default = color_default
+        self.radius_function_type = radius_function_type
+        self.radius_default = radius_default
+        self.stroke_color = stroke_color
+        self.stroke_width = stroke_width
+        self.legend_key_shape = legend_key_shape
+
+    def get_layer_specific_variables(self, options):
+        """Update map template variables specific to graduated circle visual"""
+        options.update(dict(
+            geojson_data=json.dumps(self.data, ensure_ascii=False),
+            colorProperty=self.color_property,
+            colorStops=self.color_stops,
+            colorType=self.color_function_type,
+            radiusType=self.radius_function_type,
+            defaultColor=self.color_default,
+            defaultRadius=self.radius_default,
+            radiusProperty=self.radius_property,
+            radiusStops=self.radius_stops,
+            strokeWidth=self.stroke_width,
+            strokeColor=self.stroke_color,
+            labelColor=self.label_color,
+            labelSize=self.label_size,
+            labelHaloColor=self.label_halo_color,
+            labelHaloWidth=self.label_halo_width
+        ))
+
+
+class HeatmapLayer(AbstractLayer):
+    """Create a heatmap layer"""
+
+    def __init__(self,
+                 data,
+                 label_property=None,
+                 weight_property=None,
+                 weight_stops=None,
+                 color_stops=None,
+                 radius_stops=None,
+                 intensity_stops=None,
+                 *args,
+                 **kwargs):
+        """Create a heatmap layer
+
+        :param weight_property: property to determine heatmap weight. EX. "population"
+        :param weight_stops: stops to determine heatmap weight.  EX. [[10, 0], [100, 1]]
+        :param color_stops: stops to determine heatmap color.  EX. [[0, "red"], [0.5, "blue"], [1, "green"]]
+        :param radius_stops: stops to determine heatmap radius based on zoom.  EX: [[0, 1], [12, 30]]
+        :param intensity_stops: stops to determine the heatmap intensity based on zoom. EX: [[0, 0.1], [20, 5]]
+
+        """
+        super(HeatmapLayer, self).__init__(data, *args, **kwargs)
+
+        self.template = 'heatmap'
+        self.label_property = label_property
+        self.weight_property = weight_property
+        self.weight_stops = weight_stops
+        if color_stops:
+            # Make the first color stop in a heatmap have opacity 0 for good visual effect
+            self.color_stops = [[0.00001, 'rgba(0,0,0,0)']] + color_stops
+        self.radius_stops = radius_stops
+        self.intensity_stops = intensity_stops
+
+    def get_layer_specific_variables(self, options):
+        """Update map template variables specific to heatmap visual"""
+        options.update(dict(
+            geojson_data=json.dumps(self.data, ensure_ascii=False),
```

Comment:
here and in some other places, `json.dumps` potentially can raise Exception that some objects are not JSON serializable. So, I think it would be useful to pass a custom `serializer` class into Layer's init somehow.
And in some corresponding/or similar situations be able to assign custom `serializer`

Reply:
Hmm interesting, Ill see if I can figure something out. Any suggestions are welcome!

Reply:
Some suggestion
maybe, `AbstractLayer` can receive two types of `data` in `__init__` 
- dict
- str (in this case, user is responsible for serialization as he wants)
- `__init__` can receive additional kwarg `data_serializator_class` which (if not `None`) will be used in `json.dumps(data, cls=data_serializator_class)`
- `json.dumps` can be moved to separate function of base class, like `def serialize_data` which will check whether data has type `str` or `dict` and make serialization accordingly

Reply:
this approach will also give user ability to override `def serialize_data` for example, to use another serialization library (for ex. `nssjson` or `ultrajson`) if he wants to..